internal/exec/stages/disks/luks.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2020 Red Hat
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The storage stage is responsible for partitioning disks, creating RAID
+// arrays, formatting partitions, writing files, writing systemd units, and
+// writing network units.
+// createRaids creates the raid arrays described in config.Storage.Raid.
+
+package disks
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/coreos/ignition/v2/config/util"
+	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
+	"github.com/coreos/ignition/v2/internal/distro"
+	execUtil "github.com/coreos/ignition/v2/internal/exec/util"
+)
+
+var defaultKeyFilePath = "/sysroot/etc/luks/%s"
+
+type Tang struct {
+	URL        string `json:"url"`
+	Thumbprint string `json:"thp,omitempty"`
+}
+
+type Pin struct {
+	Tpm  bool   `json:"tpm"`
+	Tang []Tang `json:"tang,omitempty"`
+}
+
+func (p Pin) MarshalJSON() ([]byte, error) {
+	if p.Tpm {
+		return json.Marshal(&struct {
+			Tang []Tang   `json:"tang,omitempty"`
+			Tpm  struct{} `json:"tpm2"`
+		}{
+			Tang: p.Tang,
+			Tpm:  struct{}{},
+		})
+	}
+	return json.Marshal(&struct {
+		Tang []Tang `json:"tang"`
+	}{
+		Tang: p.Tang,
+	})
+}
+
+type Clevis struct {
+	Pins      Pin `json:"pins"`
+	Threshold int `json:"t"`
+}
+
+func (s *stage) createLuks(config types.Config) error {
+	if len(config.Storage.Luks) == 0 {
+		return nil
+	}
+
+	for _, luks := range config.Storage.Luks {
+		// track whether Ignition creates the KeyFile
+		// so that it can be removed creation
+		var ignitionCreatedKeyFile bool
+		keyFilePath := filepath.Join("/sysroot", execUtil.LuksKeyFilePath, luks.Name)
```

Comment:
At least in the `ignitionCreatedKeyFile = true` path: Since the point of this key is to be injected into the TPM, let's store it in `/run` to ensure it never persists to disk.  We could do even better and keep it as an unlinked tempfile that we pass to `cryptsetup` as a file descriptor too.